Task framework

HakonRydland opened the pull request:
## What type of PR is this.
1. [ ] Bug
2. [x] Change
3. [ ] Enhancement

### What have you changed and why?
Information:
    New task framework (WIP)

### Please specify which Issue this PR Resolves.
closes #XXXX

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [ ] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: 

********************************************************
Notes:

- 03.March
System design open for major changes to correct shortcomings
Only a test task exists and the old task should be converted over to act as a temporary source of tasks until new task can be written (new task will be a resource heavy process)

HakonRydland committed:
First itteration
Test task is working with the system, no obvious errors

HakonRydland committed:
Renamed component to 'Tasks'

HakonRydland committed:
added task saving and load from save

HakonRydland committed:
added support for chained tasks

Bob-Murphy commented on A3A/addons/Tasks/Core/fn_getSettings.sqf:
```diff
@@ -0,0 +1,22 @@
+//CBA settings
+if (isClass (configfile >> "CBA_Extended_EventHandlers")) then {
+
+    [QGVAR(TaskLingerTime), "SLIDER", ["Task linger time", "Time in minutes a task remains in the task list after compleation"], ["Antistasi","Tasks"], [1, 15, 3, 0], true, {
```
completion

Bob-Murphy commented on A3A/addons/Tasks/Core/fn_getSettings.sqf:
```diff
@@ -0,0 +1,22 @@
+        GVAR(Settings) set ["TaskLingerTime",round _this * 60];
+    }] call CBA_fnc_addSetting;
+
+    [QGVAR(MaxTestMissions), "SLIDER", ["Max test missions", "maximum amount of test missions that can be active at the same time"], ["Antistasi","Tasks"], [0, 10, 3, 0], true, {
```
maximum --> Maximum

Bob-Murphy commented on A3A/addons/Tasks/Core/fn_requestTask.sqf:
```diff
@@ -0,0 +1,69 @@
+    //chained tasks
+    if (isClass (_cfg/"chain")) then {
+        private _chain = getText (_cfg/"chain"/"name");
+        if (_chain isEqualTo "") then { Error_1("Missconfigured task %1, lacking name of chain", _x); continueWith false; };
```
Misconfigured

Bob-Murphy commented on A3A/addons/Tasks/Core/fn_requestTask.sqf:
```diff
@@ -0,0 +1,69 @@
+
+        private _stage = GVAR(ChainStates) getOrDefault [_chain, 1];
+        if (_stage isNotEqualTo _requiredStage) then {
+            Debug_1("%1 Chain not at the required stage for task %2 | Current: %3 | Required: %4", _chain, _x, _stage, _requiredStage);
```
chain lowercase as not start of sentence

Bob-Murphy reviewed (changes requested):
typos

HakonRydland committed:
typos

HakonRydland committed:
fixed: passing task hm to all function calls in runtask

HakonRydland committed:
- improved saving efficiency
- added task weights
- cached task params

killerswin2 commented:
I don't understand why you are making tasks a separate pbo. The task are deep ingrained into core, so why not place in core?

killerswin2 commented on A3A/addons/core/functions/Save/fn_loadStat.sqf:
```diff
@@ -42,7 +42,7 @@ private _specialVarLoads = [
     "idleassets","chopForest","weather","killZones","jna_dataList","controlsSDK","mrkCSAT","nextTick",
     "bombRuns","wurzelGarrison","aggressionOccupants", "aggressionInvaders",
-    "countCA", "attackCountdownInvaders", "testingTimerIsActive", "version", "HR_Garage","A3A_fuelAmountleftArray"
+    "countCA", "attackCountdownInvaders", "testingTimerIsActive", "version", "HR_Garage","A3A_fuelAmountleftArray", QEGVAR(Tasks,ActiveTasks)
```
no macros in loadStat please

HakonRydland replied:
theres no reson not to use macros

killerswin2 commented on A3A/addons/Tasks/Tasks/fn_testTask.sqf:
```diff
@@ -0,0 +1,63 @@
+_this set ["description", "Test task description"];
+_this set ["destination", [allPlayers#0, true]];
+
+_this set ["Constructor", { // Type: code | Required | Constructor to run at start of task
```
what are we doing here?

HakonRydland replied:
how is this a change request?

killerswin2 replied:
yeah dismiss the review

HakonRydland commented on A3A/addons/Tasks/Core/fn_getSettings.sqf:
```diff
@@ -0,0 +1,22 @@
+if (isClass (configfile >> "CBA_Extended_EventHandlers")) then {
+
+    [QGVAR(TaskLingerTime), "SLIDER", ["Task linger time", "Time in minutes a task remains in the task list after completion"], ["Antistasi","Tasks"], [1, 15, 3, 0], true, {
+        GVAR(Settings) set ["TaskLingerTime",round _this * 60];
```
seconds

HakonRydland committed:
fixed syntax errors whit requested changes

HakonRydland committed:
corrected weigthed array value weight order

HakonRydland committed:
streamlined weigth based randomisation

HakonRydland committed:
ported legacy task as official

HakonRydland committed:
Merge branch 'unstable' into TaskFramework

HakonRydland committed:
added punishment for failing mission... whops

HakonRydland committed:
added support for addon dependancies for missions
(uses standard requiredAddons syntax in config)

HakonRydland committed:
Merge branch 'unstable' into TaskFramework

HakonRydland committed:
unified timelimit to one line

HakonRydland committed:
ported specOp task to new framework

killerswin2 commented on A3A/addons/Tasks/Params/LegacyParams/fn_AS_specOP_p.sqf:
```diff
@@ -0,0 +1,13 @@
+
+private _possibleMarkers = [];
+private _controlsX = [controlsX] FUNC(nearHostileMarkers);
```
call FUNC(nearHostileMarkers);

HakonRydland replied:
good catch

killerswin2 commented on A3A/addons/Tasks/Params/LegacyParams/fn_AS_specOP_p.sqf:
```diff
@@ -0,0 +1,13 @@
+
+private _possibleMarkers = [];
+private _controlsX = [controlsX] FUNC(nearHostileMarkers);
+private _nearbyFriendlyMarkers = FUNC(nearFriendlyMarkers);
```
call FUNC(nearFriendlyMarkers);

killerswin2 commented on A3A/addons/Tasks/Core/fn_getSettings.sqf:
```diff
@@ -0,0 +1,22 @@
+if !("MaxTestMissions" in GVAR(Settings)) then { GVAR(Settings) set ["MaxTestMissions", 3] };
+
+//CBA settings
+if (isClass (configfile >> "CBA_Extended_EventHandlers")) then {
```
we should make this A3A_hasCBA

killerswin2 replied:
more than one place to use this

HakonRydland replied:
this check (or similar check) would be needed anytime we have a cba dependent optional feature

HakonRydland committed:
added missing calls in new task params getter

Lazejun commented on A3A/addons/Tasks/Params/LegacyParams/fn_AS_Officail_p.sqf:
```diff
@@ -0,0 +1,24 @@
+*/
+#include "..\..\script_component.hpp"
+FIX_LINE_NUMBERS()
+private _possibleMarkers = [citiesX] call FUNC(nearHostileMarkers);
```
You kinda mixed Official and the Traitor Mission.

Traitor was Towns,
Official was Airbases.

I dont think its a bad idea to move them to Towns, just weird since the Goal was to port the old ones

Lazejun replied:
Also very good filename type

"fn_AS_Officail_p.sqf" should have been "fn_AS_Official_p.sqf"

HakonRydland committed:
- corrected typo in filename
- corrected incorrect site pool for as_official

HakonRydland committed:
Merge branch 'unstable' into TaskFramework

HakonRydland committed:
fixed required addons check

HakonRydland committed:
corrected max task of cat default when no setting has been mae for it (deafults to 1)

HakonRydland commented on A3A/addons/Tasks/Core/fn_genTaskID.sqf:
```diff
@@ -0,0 +1,6 @@
+FIX_LINE_NUMBERS()
+if (!isServer) exitWith {};
+
+GVAR(TaskID) = GVAR(TaskID) + 1;
```
missions can have more than one of each category or even same mission, as long as the setting for max tasks of the type is not blocking it, example: we would not want multiple defend petros missions running at the same time, default will be 1 per cat

HakonRydland replied:
renaming has also been done

HakonRydland committed:
changed naming of genTaskID -> genTaskUID

HakonRydland committed:
fixed bad arguments for requiredAddons check

HakonRydland committed:
fixed bad arguments for requiredAddons check

HakonRydland committed:
corrected inverted if statement

HakonRydland committed:
corrected inverted if statement

HakonRydland committed:
added missing includes

HakonRydland committed:
fixed wrong oop

copy-on-meow committed:
Ported CON_Outpost to new Framework

copy-on-meow committed:
Fixed array concatenation of non-array type

copy-on-meow committed:
Code style and formatting improvements

HakonRydland committed:
Merge pull request #3 from Jul1aK0wal1k/tasks/CON_Outpost

Migrated Conquest mission to new mission framework

jaj22 commented on A3A/addons/Tasks/Core/fn_requestTask.sqf:
```diff
@@ -0,0 +1,94 @@
+
+//Task loading from save, format: TaskName, Params, stage, task version
+if (_taskTypes isEqualTypeParams ["",[],0,0]) exitWith {
+    private _cfg = TASK_CFG/(_taskTypes#0);
```
Typo? I don't see a TASK_CFG.

HakonRydland replied:
ln8

HakonRydland replied:
ah now i see it, corrected

HakonRydland committed:
corrected typo in macro

jaj22 commented on A3A/addons/Tasks/Core/fn_runTask.sqf:
```diff
@@ -0,0 +1,95 @@
+
+_taskHM set ["Rewards",[]];
+{
+    _taskHM set ["__CurrentStage", (_taskHM get "__CurrentStage") +1]; //increment stage count (for saving/loading)
```
Wouldn't this cause currentStage to be saved as 2 when a task is saved during its first stage?

HakonRydland replied:
good catch, moved to after stage condition is meet

HakonRydland committed:
moved stage counter increment to after stage condition is meet

HakonRydland committed:
Merge remote-tracking branch 'antistasi/unstable' into TaskFramework

HakonRydland committed:
fixed param getter bug for two legacy missions

HakonRydland committed:
update naming convention to agreed upon standard (per meeting)

HakonRydland committed:
Merge branch 'unstable' into TaskFramework

HakonRydland committed:
Merge branch 'unstable' into TaskFramework

jaj22 committed:
Fix missing task marker for CON_Outpost plus incorrect conversion of marker-type destinations

killerswin2 committed:
ported LOG_Supplies with some changes

HakonRydland committed:
Merge pull request #4 from jaj22/fix-task-destination-bugs

Fix some task destination bugs

HakonRydland committed:
Merge branch 'unstable' into TaskFramework

HakonRydland committed:
Merge branch 'TaskFramework' of https://github.com/HakonRydland/A3-Antistasi into TaskFramework

killerswin2 committed:
ported log_ammo

rautamiekka commented on A3A/addons/Tasks/Params/LegacyParams/fn_LOG_Ammo_p.sqf:
```diff
@@ -0,0 +1,17 @@
+	Returns: array of markers within max mission distance and is not rebel.
+	*/
+	params ["_Markers"];
+	_Markers = _Markers select {(getMarkerPos _x distance2D getMarkerPos respawnTeamPlayer < distanceMission) && (sidesX getVariable [_x,sideUnknown] != teamPlayer)};
+	_Markers
```
```suggestion
	_Markers select {(getMarkerPos _x distance2D getMarkerPos respawnTeamPlayer < distanceMission) && (sidesX getVariable [_x,sideUnknown] != teamPlayer)}
```
No need to save privately if you immediately return.

rautamiekka commented on A3A/addons/Tasks/Params/LegacyParams/fn_LOG_Ammo_p.sqf:
```diff
@@ -0,0 +1,17 @@
+};
+
+//Add unspawned outposts for ammo trucks, and seaports for salvage
+_possibleMarkers = [ outposts] call _findIfNearAndHostile;
```
```suggestion
_possibleMarkers = [outposts] call _findIfNearAndHostile;
```

rautamiekka commented on A3A/addons/Tasks/Params/LegacyParams/fn_LOG_Ammo_p.sqf:
```diff
@@ -0,0 +1,17 @@
+_possibleMarkers = [ outposts] call _findIfNearAndHostile;
+_possibleMarkers = _possibleMarkers select {(_x in seaports) or (spawner getVariable _x != 0)};
+
+if ( _possibleMarkers isEqualTo []) exitWith {false};
```
```suggestion
if (_possibleMarkers isEqualTo []) exitWith {false};
```

rautamiekka commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+
+
+private _nameDest = [_marker] call A3A_fnc_localizar;
+private _timeLimit = if (_this get "_isDifficult") then {30} else {60};
```
```suggestion
private _timeLimit = [60, 30] select (_this get "_isDifficult");
```

HakonRydland replied:
select is slower than an if statement

rautamiekka replied:
Weird, you'd think there would be no point in the feature in such case.

rautamiekka commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+_this set ["_road", _road];
+_this set ["_nameDest", _nameDest];
+/*
+	Note: the task sets all feedback to players automaticly based on the info in the task hm passed to the task
```
```suggestion
	Note: the task sets all feedback to players automatically based on the info in the task hm passed to the task
```

rautamiekka commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+_this set ["_nameDest", _nameDest];
+/*
+	Note: the task sets all feedback to players automaticly based on the info in the task hm passed to the task
+	the modifyable values are:
```
```suggestion
	the modifiable values are:
```

rautamiekka commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+		["required", true], //Type: bool | Optional | if the task needs the stage to succeed
+		["reward", { //Type: code | Optional | the reward given for completing the stage
+			Info("Stage 1 reward called");
+			private _bonus = if (_this get "_isDifficult") then {2} else {1};
+			private _positionX =(_this get "destination");
+			private _markerX =(_this get "marker");
+			private _side = _this get "_side";
```
```suggestion
			private _bonus = [1, 2] select (_this get "_isDifficult");
			private _positionX = (_this get "destination");
			private _markerX = (_this get "marker");
			private _side = (_this get "_side");
```
The last for consistency, otherwise the parens ain't needed.

rautamiekka commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+				[5*_bonus * tierWar,theBoss] call A3A_fnc_playerScoreAdd;
+				[0,200 * _bonus * tierWar] remoteExec ["A3A_fnc_resourcesFIA",2];
+			};
+			if((_this get "state") isNotEqualTo "SUCCEEDED") then
+			{
+				[-10*_bonus,theBoss] call A3A_fnc_playerScoreAdd;
```
```suggestion
			if ((_this get "state") isNotEqualTo "SUCCEEDED") then
			{
				[-10 * _bonus,theBoss] call A3A_fnc_playerScoreAdd;
```
For consistency.

rautamiekka commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Supplies.sqf:
```diff
@@ -84,9 +84,8 @@ _stages = [
-            Info("Stage 1 reward called");
             private _bonus = if (_this get "_isDifficult") then {2} else {1};
-            private _markerX =(_this get "destination");
+            private _positionX =(_this get "destination");
```
```suggestion
            private _positionX = (_this get "destination");
```
For consistency.

rautamiekka commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Supplies.sqf:
```diff
@@ -98,7 +97,7 @@ _stages = [
 			    {[10*_bonus * tierWar,_x] call A3A_fnc_playerScoreAdd} forEach _playersInRange;
 			    [5*_bonus * tierWar,theBoss] call A3A_fnc_playerScoreAdd;
-			    [-15*_bonus,15*_bonus,_markerX] remoteExec ["A3A_fnc_citySupportChange",2];
+			    [-15*_bonus,15*_bonus,_positionX] remoteExec ["A3A_fnc_citySupportChange",2];
```
```suggestion
			    [-15 * _bonus,15 * _bonus,_positionX] remoteExec ["A3A_fnc_citySupportChange",2];
```
For consistency.

HakonRydland committed:
Update A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

HakonRydland committed:
Update A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

HakonRydland committed:
Update A3A/addons/Tasks/Params/LegacyParams/fn_LOG_Ammo_p.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

HakonRydland committed:
Update A3A/addons/Tasks/Params/LegacyParams/fn_LOG_Ammo_p.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

HakonRydland committed:
Update A3A/addons/Tasks/Params/LegacyParams/fn_LOG_Ammo_p.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

HakonRydland committed:
Update A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

HakonRydland committed:
Update A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Supplies.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

HakonRydland committed:
Update A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Supplies.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

killerswin2 commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_AS_Official.sqf:
```diff
@@ -0,0 +1,92 @@
+            _group selectLeader _official;
+
+            //patrol behaviour to be replaced by vcom
+            [leader _group, _this get "marker", "SAFE", "SPAWNED", "NOVEH", "NOFOLLOW"] execVM EQPATHTOFOLDER(core,scripts\UPSMON.sqf);
```
```suggestion
            [leader _group, _this get "marker", "SAFE", "SPAWNED", "NOVEH", "NOFOLLOW"] spawn UPSMON_fnc_UPSMON;
```

killerswin2 replied:
need to update to after upsmon move

HakonRydland committed:
Update A3A/addons/Tasks/Tasks/LegacyTasks/fn_AS_Official.sqf

Co-authored-by: Ellis Nielsen <thecakebecamethelie@yahoo.com>

Bob-Murphy commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+			};
+			if (((_this get "state") isEqualTo "SUCCEEDED") and (not alive (_this get "truckX"))) then
+			{
+				[petros,"hint",(localize "STR_antistasi_LTasks_LOG_Ammo_destroyed_description"), "Logistics Mission"] remoteExec ["A3A_fnc_commsMP",[teamPlayer,civilian]];
```
"Logistics Mission"
Please localize

Bob-Murphy commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+			// OKAY TASK IS GOOD
+			if (((_this get "state") isEqualTo "SUCCEEDED") and (alive (_this get "truckX"))) then
+			{
+				[petros,"hint",(localize "STR_antistasi_LTasks_LOG_Ammo_captured"), "Logistics Mission"] remoteExec ["A3A_fnc_commsMP",[teamPlayer,civilian]];
```
"Logistics Mission"
Please localize

Bob-Murphy commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Ammo.sqf:
```diff
@@ -0,0 +1,168 @@
+				private _owningSide = (_vehicle getVariable "originalSide");		// set by AIVEHinit
+		
+				if (_unit getVariable ["spawner",false]) then {
+					["TaskFailed", ["", format ["Ammotruck Stolen in an %1",(_vehicle getVariable ["ammoTruckLocation", ""])]]] remoteExec ["BIS_fnc_showNotification",_owningSide];
```
"Ammotruck Stolen in an %1"
Please localize

Bob-Murphy commented on A3A/addons/Tasks/Tasks/LegacyTasks/fn_LOG_Supplies.sqf:
```diff
@@ -0,0 +1,120 @@
+            // OKAY TASK IS GOOD
+			if ((_this get "state") isEqualTo "SUCCEEDED") then
+			{
+                [petros,"hint",(localize "STR_antistasi_LTasks_LOG_Supplies_supplies_delivered"), "Logistics Mission"] remoteExec ["A3A_fnc_commsMP",[teamPlayer,civilian]];
```
"Logistics Mission"
Please localize

Bob-Murphy commented on A3A/addons/Tasks/stringtable.xml:
```diff
@@ -0,0 +1,73 @@
+      </Key>
+      <Key ID="STR_antistasi_LTasks_AS_specOP_description">
+      <!-- %1 = Faction name | %2 = Location name | %3 = time in format H:M -->
+        <Original>We have spotted a %1 SpecOp team patrolling around a %2. Ambush them and we will have one less problem. Do this before %3. Be careful, they are tough boys.</Original>
```
Just double-checking, does it makes to have the `a` before the %2?
It is stating something like "ourpost, airport, resource" or for example a city name?
Maybe `around the` would be more fitting as for `outpost/airport` it would be `an outpost / an airfield`.

Bob-Murphy commented on A3A/addons/Tasks/stringtable.xml:
```diff
@@ -0,0 +1,73 @@
+      </Key> 
+      <!-- %1 = Location name | %2 = time in format H:M -->
+      <Key ID="STR_antistasi_LTasks_CON_Outpost_resource_description">
+        <Original>A %1 would be a fine addition to our cause. Go there and capture it before %2.</Original>
```
Same here, what exactly is the location name? Does the `A` make sense?

Bob-Murphy commented on A3A/addons/Tasks/stringtable.xml:
```diff
@@ -0,0 +1,73 @@
+      </Key> 
+      <!-- %1 = Location name | %2 = time in format H:M -->
+      <Key ID="STR_antistasi_LTasks_CON_Outpost_outpost_description">
+        <Original>A %1 is disturbing our operations in the area. Go there and capture it before %2.</Original>
```
Same as above

Bob-Murphy commented on A3A/addons/Tasks/stringtable.xml:
```diff
@@ -0,0 +1,73 @@
+        <Original>City Supplies</Original>
+      </Key>
+      <Key ID="STR_antistasi_LTasks_LOG_Supplies_description">
+        <Original>%1 population is in need of supplies. We may improve our relationship with that city if we are the ones who provide them. I have placed a crate with supplies near our HQ. Deliver the crate to %1 city center, hold it there for %3 minutes and it's done. Do this before %2.</Original>
```
You mention `city center` but we mostly deliver to villages.
Maybe something like `Deliver the crate to the center of %1` makes more sense.

Bob-Murphy commented on A3A/addons/Tasks/stringtable.xml:
```diff
@@ -0,0 +1,73 @@
+        <Original>%1 population is in need of supplies. We may improve our relationship with that city if we are the ones who provide them. I have placed a crate with supplies near our HQ. Deliver the crate to %1 city center, hold it there for %3 minutes and it's done. Do this before %2.</Original>
+      </Key>
+      <Key ID="STR_antistasi_LTasks_LOG_Supplies_distributing_supplies">
+        <Original>Keep distributing supplies for %1 more seconds.</Original>
```
You are not actively distributing the supplies but more guarding them.
`Keep guarding the supplies for %1 more seconds.`

Bob-Murphy commented on A3A/addons/Tasks/stringtable.xml:
```diff
@@ -0,0 +1,73 @@
+        <Original>Steal or Destroy Ammotruck</Original>
+      </Key>
+      <Key ID="STR_antistasi_LTasks_LOG_Ammo_description">
+        <Original>We've spotted an Ammotruck in an %1. Go there and destroy or steal it before %2.%2.</Original>
```
duplication of `%2`?

Bob-Murphy reviewed (changes requested):
as noted

HakonRydland committed:
- updated the task runner to correctly give out rewards and punishments
- added comments to make the proccess more readable
- minor code readability improvements

Tiny-DM commented:
An expanded version of the task state machine was added in #3667. It is finally time to close another historical piece.
